### DISCUSSION TITLE
More Python-like loops

### DISCUSSION THREAD

**Discussion by richlander** [dotnet/csharplang#6485] [Language Ideas]
I really like Python based loops, at least as I see them written. They are terse and intuitive. C# `for` loops are terse-ish, but antiquated/unintutive. C# `foreach` loops are bulky, but intuitive.

From https://github.com/adafruit/Adafruit_CircuitPython_IS31FL3731/blob/main/examples/is31fl3731_rgbmatrix5x5_rainbow.py#L100-L102:

```python
    for _ in range(100):
        for y in range(0, 5):
            for x in range(0, 5):
```

Gotta love that discard!

I shared this with @MadsTorgersen. He wrote the following code as the best we have today:

```csharp
using static System.Linq.Enumerable;

foreach (var _ in Range(0, 100))
{
    foreach (var y in Range(0, 5))
    {
        foreach (var x in Range(0, 5))
        {
            Console.WriteLine($"{x}  {y}");
        }
    }
}
```

That's actually pretty good. I had forgotten about static imports.

Ideally, the following would work:

```csharp
foreach (_ in 0..5)
{
    foreach (var y in 0..5)
    {
        foreach (var x in 0..5)
        {
            Console.WriteLine($"{x}  {y}");
        }
    }
}
```

If we had that, I'd say that C# loops were at least as good as (if not better than) Python loops.

**Comment by FaustVX** [comment]
You could write an `GetEnumerator()` extension methos on `Range` to allow this.

**Reply by richlander** [reply]
Sure. I'm looking for a change we could offer generally.

**Reply by RikkiGibson** [reply]
I do think it's good that people can write this extension, but I also really feel that this is something that should come "in the box".

**Reply by Zodt** [reply]
A good example of how it should look like

```csharp
// C#9
using System;
using System.Collections.Generic;

foreach (var (value, index) in 5..10)
{
    Console.WriteLine($"Value = {value.ToString()}, Index = {index.ToString()}");
}

public static class RangeExtensions
{
    public static IEnumerator<ValueTuple<int, Index>> GetEnumerator(this Range range)
    {
        if (range.Start.Value == range.End.Value)
            throw new ArgumentOutOfRangeException(nameof(range), 
                "The initial value of the Range should not be equal to the final one");

        int index = default;
        var enumerator = new RangeEnumerator(range);
        for (var value = range.Start.Value; enumerator.Condition(value); enumerator.Expression(ref value))
            yield return (value, index++);
    }

    private readonly struct RangeEnumerator
    {
        private readonly Range _range;
        public RangeEnumerator(Range range) => _range = range;
        private T GetResult<T>(T result, T revertResult) => _range.Start.Value > _range.End.Value ? revertResult : result;
        internal void Expression(ref int value) => value = GetResult(++value, --value);
        internal bool Condition(int value) => GetResult(value < _range.End.Value, value > _range.End.Value);
    }
}   
```

**Comment by HaloFour** [comment]
I recall that the team had went back and forth over the enumerability of ranges a number of times.  I believe there was concern around confusion with inclusivity and open ranges, the latter of which would effectively have to throw.  The language could theoretically treat range literally differently in this case, although it would seemingly still always have to favor an extension method unless the team was willing to accept breaking changes here.

As for discards, that seems like such a minor thing, but at least it's not legal syntax today so it could be implemented without ambiguity.

**Reply by HaloFour** [reply]
Honestly I think the best syntactic improvement to that C# version is simply not to use Allman braces.  Plus, with the extension method, which you could import globally, you've got:

```cs
foreach (var _ in 0..100)
    foreach (var y in 0..5)
        foreach (var x in 0..5)
            Console.WriteLine($"{x}  {y}");
```

**Reply by richlander** [reply]
Nice!

**Reply by RikkiGibson** [reply]
I would like to see enumerable ranges become a thing, i.e. ranges whose indexes are always "from start".

You could define a conversion from range literal to EnumerableRange, and use such a conversion in constructs like `foreach` where an enumerable thing is needed.

```cs
foreach (var i in 1..10) { } // ok
foreach (var in 1..^1) { } // error: can't convert literal to EnumerableRange

EnumerableRange er = 1..10;
int[] arr = ...;
int[] arr1 = arr[er]; // ok: we'd have an implicit conversion from EnumerableRange to Range
```

I haven't looked closely at the inclusivity issue, though.

**Reply by qrli** [reply]
It would be tricky to loop backward down to 0.
```cs
foreach (var i in 10..0) { } // from 10 to 1, but not to 0
foreach (var i in (9..-1)) { } // Argument out of range exception

foreach (var i in (0..10).Reverse()) { } // creates a special EnumerableRange, which cannot convert to Range...
```

**Reply by jnm2** [reply]
Besides the fact that C# ranges are end-exclusive and the question of whether to count backwards or do nothing when the end is smaller than the start, there's also the fact that either side of a range can be a from-end index (e.g. `^4`) which would have to be an error at runtime. Simple cases could be caught at compile time but not all.

**Reply by HaloFour** [reply]
This is probably why Python has different syntaxes for slicing vs. ranges.  By targeting slicing with ranges syntactic and behavior choices were made that would make them awkward to use with enumeration.  Of course the compiler could decide to change the rules when it comes to a range expression in `foreach` and not rely on the behavior of `System.Index` or `System.Range`.  I'm not opposed to the conversation, I just don't think it's a simple one.

**Reply by jnm2** [reply]
If the rules were changed when the expression appeared in foreach, then the obvious ways of extracting a variable or method to hold or return the range would change the behavior.

**Reply by iam3yal** [reply]
I know that in most cases it's desired to use existing types but they can always add another like `InclusiveRange` for this purpose and then add APIs for conversions or something so at least if you need to convert from/to you do this explicitly.
